lib/metadata/include/irods/metadata.hpp

```diff
@@ -0,0 +1,163 @@
+                  std::string units;
+              };
+        bool operator==(const avu& _lhs, const avu& _rhs) noexcept;
+
+namespace {
+        /// \brief Tokens used for representing entity type for rxModAVUMetadata
+        /// \since 4.2.9
+        namespace tokens {
+            const std::string collection{"-C"};
+            const std::string data_object{"-d"};
+            const std::string user{"-u"};
+            const std::string resource{"-R"};
+        }
+
+
+        /// \brief Map which converts an entity type to a string representation
+        /// \since 4.2.9
+        const std::map<entity_type, std::string> type_to_string {
+            {entity_type::collection,  "collection"},
+            {entity_type::data_object, "data_object"},
+            {entity_type::user,        "user"},
+            {entity_type::resource,    "resource"}
+        };
+
+        /// \brief Map which converts an entity type to a rxModAVUMetadata token
+        /// \since 4.2.9
+        const std::map<entity_type, std::string> type_to_token {
+            {entity_type::collection,  tokens::collection},
+            {entity_type::data_object, tokens::data_object},
+            {entity_type::user,        tokens::user},
+            {entity_type::resource,    tokens::resource}
+        };
+
+        /// \brief Map which converts a rxModAVUMetadata token to an entity type
+        /// \since 4.2.9
+        const std::map<std::string, entity_type> token_to_type {
+            {tokens::collection,  entity_type::collection},
+            {tokens::data_object, entity_type::data_object},
+            {tokens::user,        entity_type::user},
+            {tokens::resource,    entity_type::resource}
+        };
+
+        /// \brief convenience function for invoking rxModAVUMetadata
+        /// \since 4.2.9
+        int mod_avu_meta(rxComm_t& _comm, const avu& _md, const avu& _dmd, entity_type _et, const std::string& _tgt, const std::string& _op)
+        {
+            modAVUMetadataInp_t inp{
+                const_cast<char*>(_op.c_str()),
+                const_cast<char*>(type_to_token.at(_et).c_str()),
+                const_cast<char*>(_tgt.c_str()),
+                const_cast<char*>(_md.attribute.c_str()),
+                const_cast<char*>(_md.value.c_str()),
+                const_cast<char*>(_md.units.c_str()),
+                const_cast<char*>(_dmd.attribute.c_str()),
+                const_cast<char*>(_dmd.value.c_str()),
+                const_cast<char*>(_dmd.units.c_str()),
+                nullptr, // unused arg9
+                KeyValPair{}
+            };
+
+            return rxModAVUMetadata(&_comm, &inp);
+        } // mod_avu_meta
+
+        template<typename M_T, typename T_T>
+        void throw_if_invalid(const M_T& _m, const T_T& _t)
+        {
+            if(_m.find(_t) == _m.end()) {
+                THROW(SYS_INVALID_INPUT_PARAM, "token or entity not found");
+            }
+        }
+
+    } // namespace
+
```

Comment:
Let's tighten up the consecutive empty lines.

Squash them into a single empty line.